x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/ProcessPipes.java

```diff
@@ -118,19 +119,35 @@ public void addArgs(List<String> command) {
     }
 
     /**
-     * Connect the pipes created by the C++ process.  This must be called after the corresponding C++ process has been started.
+     * Connect the log pipe created by the C++ process.  The must be connected before any other pipes <em>and a thread must be
+     * started to read from it</em>so that there is no risk of messages logged in between creation of the other pipes on the C++
+     * side from blocking due to filling up the named pipe's buffer, and hence deadlocking communications between that process
+     * and this JVM.
+     * @param timeout Needs to be long enough for the C++ process perform all startup tasks that precede creation of named pipes.
+     *                There should not be very many of these, so a short timeout should be fine.  However, at least a couple of
+     *                seconds is recommended due to the vagaries of process scheduling and the way VMs can completely stall for
+     *                some hypervisor actions.
+     */
+    public void connectLogStream(Duration timeout) throws IOException {
+        logStreamHandler = new CppLogMessageHandler(jobId, namedPipeHelper.openNamedPipeInputStream(logPipeName, timeout));
+    }
+
+    /**
+     * Connect the other pipes created by the C++ process after the logging pipe has been connected.  This must be called after
+     * the corresponding C++ process has been started, and after {@link #connectLogStream}.
      * @param timeout Needs to be long enough for the C++ process perform all startup tasks that precede creation of named pipes.
      *                There should not be very many of these, so a short timeout should be fine.  However, at least a couple of
      *                seconds is recommended due to the vagaries of process scheduling and the way VMs can completely stall for
      *                some hypervisor actions.
      */
-    public void connectStreams(Duration timeout) throws IOException {
+    public void connectOtherStreams(Duration timeout) throws IOException {
+        assert logStreamHandler != null : "Must connect log stream before other streams";
+        if (logStreamHandler == null) {
```

Comment:
Why do we need both `assert` and `if` statements?
Is it because `assert`s can get optimized away?

Reply:
Yes, in production we run with assertions disabled.  So the assertion is to make CI fail instantly if the condition is violated (as it indicates a mistake in the program, not an environmental or user problem).  But if the mistake _did_ somehow make it into production then we'll get a normal exception.